### PR TITLE
Messages

### PR DESCRIPTION
**Pull request by freefal**
This PR adds the ability for the user to view his inbox, view threads, respond to threads, delete threads, and compose new threads. It also adds a button on the profile page of each user to allow messaging.

It does not include a notification system to alert a user to a new message when it arrives.

Closes https://github.com/veloce/lichobile/issues/308

### CODE REVIEW

**Comment by veloce** [comment]
Whaa, thank you very much! It looks good, I'll review it asap. Just one thing for now: could you also write views in typescript? Normally view files are just as easy to write in typescript as the rest.

**Comment by freefal** [comment]
Sure thing. I tried at one point, but the typscript compiler seemed to be throwing up on the mithril templating, but I'll give it another shot this week. I'll mark this WIP until done, but feel free to comment in the meantime.

File: project/src/js/ui/inbox/thread/threadView.tsx [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+function renderPost(post: Post, index: number, posts: Array<Post>) {
+  let postClass = 'postWrapper';
+  if (index === 0)
+    postClass += ' first';
+  if (index === posts.length-1)
+    postClass += ' last';
+  return (
+    <div className={postClass} key={post.createdAt}>
+      <span className="infos">
```
Comment (veloce) [inline project/src/js/ui/inbox/thread/threadView.tsx]: Should be a `div` since it's a block element

**Comment by freefal** [comment]
Make sense. Just pushed this change.

On Tue, Oct 11, 2016 at 4:51 AM, Vincent Velociter <notifications@github.com

**Comment by veloce** [comment]
@freefal I quickly tested the inbox in my android phone: this is really good! One thing I noticed: when you are in a thread view and want to reply, the textarea doesn't shift up when the virtual keyboard is open. It would be nice to have the view scroll up so that the textarea is a the top of the page.

**Comment by freefal** [comment]
Makes sense. Let me see if I can make that work.

On Tue, Oct 11, 2016 at 9:09 AM, Vincent Velociter <notifications@github.com

File: project/src/js/routes.js [hunk 2]
```diff
@@ -61,6 +64,10 @@ export default {
       'clock': clock,
       'editor': editor,
       'editor/:fen': editor,
+      'inbox': inbox,
+      'inbox/:id': inboxThread,
+      'inbox/new': inboxCompose,
+      'inbox/new/:id': inboxCompose,
```
Comment (veloce) [inline project/src/js/routes.js]: maybe `userId` here, to be more clear?

**Comment by freefal** [comment]
Yeah, I like that. Will change.

File: project/src/js/ui/inbox/compose/composeView.tsx [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+function composeBody(ctrl: ComposeState) {
+  return (
+    <div className="composeWrapper">
+      <form id="composeForm"
+      onsubmit={function(e: Event) {
+        e.preventDefault();
+        return ctrl.send(e.target as HTMLFormElement);
+      }}>
+        {ctrl.id() ? recipientWithName(ctrl) : recipientWithoutName()}
+        <div key="recipientError" className="errorMessage">
```
Comment (veloce) [inline project/src/js/ui/inbox/compose/composeView.tsx]: Here we have empty divs when there is no error, why not render conditionally the div as well?
You'll need to ensure every element in `<form>` has a unique key.

Idem for other error fields

File: project/src/js/ui/inbox/compose/composeView.tsx [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  );
+}
+
+function recipientWithName(ctrl: ComposeState) {
+  return (
+    <input id="recipient" type="text" className="composeInput"
+    placeholder={i18n('recipient')}
+    autocapitalize="off"
+    autocomplete="off"
+    value={ctrl.id() ? ctrl.id() : null}
```
Comment (veloce) [inline project/src/js/ui/inbox/compose/composeView.tsx]: Normally here, just `value={ctrl.id()}` is enough

File: project/src/js/ui/inbox/inboxView.tsx [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import {header } from '../shared/common';
+import layout from '../layout';
+import i18n from '../../i18n';
+import {InboxState, Thread} from './interfaces';
+
+export default function view(vnode: Mithril.Vnode<{}>) {
+  const ctrl = vnode.state as InboxState;
+  const bodyCtrl = inboxBody.bind(undefined, ctrl);
+  const footer = () => renderFooter(ctrl);
+  return layout.free(header.bind(undefined, i18n('inbox')), bodyCtrl, footer, undefined);
```
Comment (veloce) [inline project/src/js/ui/inbox/inboxView.tsx]: no `bind`

File: project/src/js/ui/inbox/inboxView.tsx [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+}
+
+function inboxBody(ctrl: InboxState) {
+  if (!ctrl.threads() || !ctrl.threads().currentPageResults) return null;
+  const threads = ctrl.threads();
+  if (threads.nbResults === 0) {
+    return (<div className="emptyInbox"> {i18n('noNewMessages')} </div>);
```
Comment (veloce) [inline project/src/js/ui/inbox/inboxView.tsx]: You don't need parenthesis for single line JSX

File: project/src/js/ui/inbox/inboxView.tsx [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+  const forwardEnabled = threads.currentPage < threads.nbPages;
+
+  return (
+    <div className="native_scroller inboxWrapper">
+      <table className="threadList">
+        <tbody>
+          {threads.currentPageResults.map(renderInboxItem)}
+        </tbody>
+      </table>
+      <div className={'navigationButtons' + (threads.nbPages <= 1 ? ' invisible' : '')}>
```
Comment (veloce) [inline project/src/js/ui/inbox/inboxView.tsx]: You already have a footer, why not put the nav buttons in it, to save space?

File: project/src/js/ui/inbox/thread/threadCtrl.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+    deleteAttempted,
+    sendResponse,
+    deleteThread
+  };
+}
+
+function sendResponse(form: HTMLFormElement) {
+  const id = (form[0] as HTMLInputElement).value;
+  const response = (form[1] as HTMLTextAreaElement).value;
+  if(!response || response === '') return;
```
Comment (veloce) [inline project/src/js/ui/inbox/thread/threadCtrl.ts]: Empty string is falsy in javascript, so you don't need the right part

File: project/src/js/ui/inbox/thread/threadCtrl.ts [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+    else {
+      redraw();
+    }
+  })
+  .catch(handleXhrError);
+}
+
+function deleteThread(id: string) {
+  xhr.deleteThread(id)
+  .then(data => {
```
Comment (veloce) [inline project/src/js/ui/inbox/thread/threadCtrl.ts]: for simple function it's better to have a onliner, and `data` param is not used:

```
.then(() => router.set('/inbox'))
```

File: project/src/js/ui/inbox/thread/threadView.tsx [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+import i18n from '../../../i18n';
+import redraw from '../../../utils/redraw';
+import { ThreadState, Post, ThreadAttrs } from '../interfaces';
+
+export default function view(vnode: Mithril.Vnode<ThreadAttrs>) {
+  const ctrl = vnode.state as ThreadState;
+  const headerCtrl = () => headerWidget(null,
+    backButton(ctrl.thread() ? ctrl.thread().name : null)
+  );
+  const bodyCtrl = threadBody.bind(undefined, ctrl);
```
Comment (veloce) [inline project/src/js/ui/inbox/thread/threadView.tsx]: no `bind`

File: project/src/js/ui/inbox/thread/threadView.tsx [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+}
+
+function deleteButton (ctrl: ThreadState) {
+  return ctrl.deleteAttempted() ? (
+    <div className="negotiation confirmDeleteDialog">
+      <button key="confirmDelete" className="fatButton confirmDelete" oncreate={h.ontapY(() => {ctrl.deleteThread(ctrl.id());})}>
+        <span className="fa fa-trash-o" />
+        Delete
+      </button>
+      <button key="cancelDelete" className="fatButton cancelDelete" oncreate={h.ontapY(() => {ctrl.deleteAttempted(false); redraw();})}>
```
Comment (veloce) [inline project/src/js/ui/inbox/thread/threadView.tsx]: I don't like onliners with two statements separated by semicolon. It's better like this:
```
<button key="cancelDelete" className="fatButton etc"
  oncreate={h.ontapY(() => {
    // do stuff
    // several lines  
 })}
/>
...
```

**Comment by freefal** [comment]
Thanks for the feedback @veolce. I've made all your suggested changes.
You're very kind to call `{ctrl.id() ? ctrl.id() : null}` a coding style
issue. ;)

On Wed, Oct 12, 2016 at 5:26 PM, Vincent Velociter <notifications@github.com

File: project/src/js/ui/inbox/compose/composeView.tsx [hunk 12]
```diff
@@ -9,11 +9,12 @@ export default function view(vnode: Mithril.Vnode<ComposeAttrs>) {
   const headerCtrl = () => headerWidget(null,
     backButton(i18n('composeMessage'))
   );
-  const bodyCtrl = composeBody.bind(undefined, ctrl);
+  const bodyCtrl = () => composeBody(ctrl);
   return layout.free(headerCtrl, bodyCtrl, undefined, undefined);
 }
 
 function composeBody(ctrl: ComposeState) {
+  console.log(ctrl.errors());
```
Comment (freefal) [inline project/src/js/ui/inbox/compose/composeView.tsx]: Woops. Fixed.

File: project/src/js/push.js [hunk 13]
```diff
@@ -18,6 +18,11 @@ function notificationOpenedCallback(data) {
           window.plugins.toast.show(
             i18n('userAcceptsYourChallenge', lightPlayerName(additionalData.userData.joiner)), 'long', 'top');
           break;
+        case 'newMessage':
+          session.refresh();
```
Comment (veloce) [inline project/src/js/push.js]: Why `session.refresh()`?

File: project/src/js/push.js [hunk 14]
```diff
@@ -18,6 +18,11 @@ function notificationOpenedCallback(data) {
...
+          window.plugins.toast.show(
+            'New mesasge from ' + lightPlayerName(additionalData.userData.sender), 'long', 'top');
```
Comment (veloce) [inline project/src/js/push.js]: Typo

**Pull request merged**